Mathlib/Combinatorics/Hypergraph/Basic.lean

```diff
@@ -0,0 +1,492 @@
+  intro e (he : e ∈ E(H))
+  simpa using he.isSubset_vertexSet
+
+lemma vertex_mem_if_mem_hyperedge {H : Hypergraph α} {e : Set α} {x : α}
+  (he : e ∈ H.hyperedgeSet) (hx : x ∈ e) : x ∈ H.vertexSet := by
+  have h1 : e ⊆ V(H) := by apply H.hyperedge_isSubset_vertexSet he
+  apply Set.mem_of_subset_of_mem h1 hx
+
+/--
+If edges `e` and `e'` have the same vertices from `G`, then they have all the same vertices.
+This could be phrased as `e = e'`, but this formulation is more useful in combination with the `ext`
+tactic.
+-/
+lemma forall_of_forall_verts {e e' : Set α} (he : e ∈ E(H)) (he' : e' ∈ E(H))
+    (h : ∀ x ∈ V(H), x ∈ e ↔ x ∈ e') : ∀ x, x ∈ e ↔ x ∈ e' :=
+  fun x ↦ ⟨fun y ↦ (h x (he.isSubset_vertexSet y)).1 y,
+  fun y ↦ (h x (he'.isSubset_vertexSet y)).2 y⟩
+
+lemma sUnion_hyperedgeSet_subset_vertexSet : Set.sUnion E(H) ⊆ V(H) := by
+  refine subset_powerset_iff.mp ?_
+  exact coe_isSubset_vertexSet_powerset
+
+/-! ## Vertex and Hyperedge Adjacency -/
+
+/--
+Predicate for adjacency. Two vertices `x` and `y` are adjacent if there is some
+hyperedge `e ∈ E(H)` where `x` and `y` are both incident on `e`.
+
+Note that we do not need to explicitly check that x, y ∈ V(H) here because a vertex that is not in
+the vertex set cannot be incident on any hyperedge.
+-/
+def Adj (H : Hypergraph α) (x : α) (y : α) : Prop :=
+  ∃ e ∈ E(H), x ∈ e ∧ y ∈ e
+
+lemma Adj.symm {H : Hypergraph α} {x y : α} (h : H.Adj x y) : H.Adj y x := by
+  unfold Adj at *
+  obtain ⟨e, he⟩ := h
+  use e
+  constructor
+  · exact he.1
+  constructor
+  · exact he.2.2
+  · exact he.2.1
+
+-- Credit: Peter Nelson, Jun Kwon
+lemma hypergraph_adj_comm (x y) : H.Adj x y ↔ H.Adj y x :=
+  ⟨.symm, .symm⟩
+
+/--
+Predicate for (hyperedge) adjacency. Analogous to `Hypergraph.Adj`, hyperedges `e` and `f` are
+adjacent if there is some vertex `x ∈ V(H)` where `x` is incident on both `e` and `f`.
+-/
+def EAdj (H : Hypergraph α) (e : Set α) (f : Set α) : Prop :=
+  e ∈ E(H) ∧ f ∈ E(H) ∧ ∃ x ∈ V(H), x ∈ e ∧ x ∈ f
+
+lemma EAdj.symm {H : Hypergraph α} {e f : Set α} (h : H.EAdj e f) : H.EAdj f e := by
+  unfold EAdj at *
+  obtain ⟨v, hv⟩ := h.2.2
+  constructor
+  · exact h.2.1
+  constructor
+  · exact h.1
+  · use v
+    constructor
+    · exact hv.1
+    constructor
+    · exact hv.2.2
+    · exact hv.2.1
+
+lemma EAdj.inter_nonempty {H : Hypergraph α} {e f : Set α} (hef : H.EAdj e f) :
+  (e ∩ f).Nonempty := by
+    unfold EAdj at *
+    have h' : ∃ x ∈ e, x ∈ f := by grind
+    apply Set.inter_nonempty.mpr h'
+
+-- Credit: Peter Nelson, Jun Kwon
+lemma hypergraph_eadj_comm (e f) : H.EAdj e f ↔ H.EAdj f e :=
+  ⟨.symm, .symm⟩
+
+/--
+Neighbors of a vertex `x` in hypergraph `H`
+
+A vertex `y` is a neighbor of vertex `x` if there exists some hyperedge `e ∈ E(H)` where `x` and
+`y` are both incident on `e`, i.e., if the two vertices are adjacent (see `Hypergraph.Adj`)
+-/
+def neighbors (H : Hypergraph α) (x : α) : Set α := {y | H.Adj x y}
```

Comment:
imo we should instead make the neighbourhood hypergraph (as I sent you), then this would just be the vertex set of that graph. This way we'd just be making a more general construction

Reply:
Yeah, that makes sense. I wasn't planning on adding the neighborhood hypergraph in this PR, so I guess for now I'll just remove this.